### PR TITLE
Add support for configurable ssh key extensions

### PR DESCRIPTION
**Pull request by lxea**
This adds support for custom ssh key extensions with syntax as specified in #7830  
```yaml
kind: role
metadata:
  name: developer
spec:
  options:
     cert_extensions:
       - type: ssh
         mode: extension
         name: login@github.com
         value: "{{ external.github_login }}"
```

### CODE REVIEW

File: lib/services/authority.go
Comment (smallinsky) [inline lib/services/authority.go:242]: add godoc

File: lib/auth/native/native_test.go
Comment (smallinsky) [inline lib/auth/native/native_test.go:206]: add validation to the `types.CertExtension`  CertExtensions type
For instance the `types.NewRole` call for   RoleSpecV4 with `CertExtension.Type: "invalid type ",`  should return an error.

**Review by smallinsky — COMMENTED** [review]
Left some comments also mentioned `github_login` trait seems to be not fetched in current github connector implementation.

File: lib/auth/auth.go
Comment (lxea) [reply lib/auth/auth.go:907]: It was just for calculating the extension. I've updated it to calculate the list of cert extensions elsewhere.

**Comment by lxea** [comment]
>  mentioned `github_login` trait seems to be not fetched in current github connector implementation.

Does a separate github_login trait need to exist instead of just external.logins ? 

https://github.com/gravitational/teleport/blob/master/lib/auth/github.go#L358 it seems to just be username here

File: api/types/types.proto
Comment (smallinsky) [inline api/types/types.proto:1512]: enum values should be have capital cases.

File: api/types/types.proto
Comment (smallinsky) [inline api/types/types.proto:1519]: enum values should have capital cases.

File: api/types/types.proto
Comment (smallinsky) [inline api/types/types.proto:1518]:
```suggestion
    // ssh is used when extending an ssh certificate.
```

File: api/types/types.proto
Comment (smallinsky) [inline api/types/types.proto:1510]:
```suggestion
    // extension represents a cert extension that may or may not be
```

File: api/types/types.proto
Comment (smallinsky) [inline api/types/types.proto:1540]:
```suggestion
    EXTENSION = 0;
```

File: api/types/extension.go
Comment (smallinsky) [inline api/types/extension.go:31]: Does this map is request. It will add additional maintenance cost and probably will quickly be outdated when a new `CertExtensionType` will be added ? Does `CertExtensionType_value` and `CertExtensionType_name` can be reused here ?

**Review by smallinsky — COMMENTED** [review]
> Does a separate github_login trait need to exist instead of just external.logins ?

Yes if the logic should work with the `"{{ external.github_login }}"` value, but I think that the ``"{{ external.logic }}"` can be used in current implementation so this is only matter of naming. 


Also Can this flow can be covered by UT ?

File: api/types/extension.go
Comment (lxea) [reply api/types/extension.go:31]: They can be used, however it makes the configuration that is available to the user quite ugly as it would require the yaml configuration for each be in all caps. They cant go out of date however as an error will be raised if the enum isn't in the maps.

**Comment by russjones** [comment]
Alex, once you're back, please add docs before you merge this.

**Comment by lxea** [comment]
> Alex, once you're back, please add docs before you merge this
> 
> Create a Guide in documentation called "Teleport with GitHub".

I've added a basic guide, although I'm not sure what else should be included, if more github specific stuff should be there for example.

**Comment by russjones** [comment]
@lxea Let's move the guide out of this PR so you can get a review on that separately and merge this PR. You might want to add @zmb3 as a reviewer as well.

Here are some docs you might want to look into. But I think you'll need the following information in the guide.

* How to configure GitHub to trust a CA.
* How to export CA from Teleport.
* How you configure your role to inject the appropriate information.
* (Maybe) how to verify that the new extensions have been included into your users certificate.

https://docs.github.com/en/organizations/managing-git-access-to-your-organizations-repositories/about-ssh-certificate-authorities
https://docs.github.com/en/organizations/managing-git-access-to-your-organizations-repositories/managing-your-organizations-ssh-certificate-authorities

**Comment by russjones** [comment]
@smallinsky @jakule Can you two do another pass?

File: lib/services/role.go
Comment (jakule) [inline lib/services/role.go:361]: Shouldn't this be logged as warning at least? Debug message starting with "Error" is a bit confusing in my opinion.

File: lib/auth/native/native.go
Comment (jakule) [inline lib/auth/native/native.go:293]: Do we have any plans to work on that?

File: api/types/extension.go
Comment (jakule) [inline api/types/extension.go:49]: https://github.com/golang/go/wiki/CodeReviewComments#error-strings
```suggestion
		return trace.Errorf("invalid certificate extension type: %q", string(b))
```

File: api/types/extension.go
Comment (jakule) [inline api/types/extension.go:36]:
```suggestion
		return nil, trace.Errorf("Invalid certificate extension type: %q", t)
```

**Review by jakule — APPROVED** [review]
Looks good. A few minor comments.

**Review by smallinsky — APPROVED** [review]
LGTM when remaining comment will be addressed

File: lib/auth/native/native.go
Comment (lxea) [reply lib/auth/native/native.go:293]: I dont know if there are concrete plans for this, it wasnt needed for the github key extensions though.

**Pull request merged**
